cloud/openstack/os_router.py

```diff
@@ -0,0 +1,138 @@
+     required: false
+     default: true
+requirements: ["shade"]
+'''
+
+EXAMPLES = '''
+# Creates a router for tenant admin
+- os_router:
+    state=present
+    name=router1
+    admin_state_up=True
+'''
+
+
+def _needs_update(router, admin_state_up):
+    """Decide if the given router needs an update.
+
+    The only attribute of the router that we allow to change is the value
+    of admin_state_up. Name changes are not supported here.
+    """
+    if router['admin_state_up'] != admin_state_up:
+        return True
+    return False
+
+def _system_state_change(module, router):
+    """Check if the system state would be changed."""
+    state = module.params['state']
+    if state == 'absent' and router:
+        return True
+    if state == 'present':
+        if not router:
+            return True
+        return _needs_update(router, module.params['admin_state_up'])
+    return False
+
+def main():
+    argument_spec = openstack_full_argument_spec(
+        name=dict(required=True),
+        admin_state_up=dict(type='bool', default=True),
+        state=dict(default='present', choices=['absent', 'present']),
+    )
+
+    module_kwargs = openstack_module_kwargs()
+    module = AnsibleModule(argument_spec,
+                           supports_check_mode=True,
+                           **module_kwargs)
+
+    if not HAS_SHADE:
+        module.fail_json(msg='shade is required for this module')
+
+    name = module.params['name']
+    admin_state_up = module.params['admin_state_up']
+    state = module.params['state']
+
+    try:
+        cloud = shade.openstack_cloud(**module.params)
+        router = cloud.get_router(name)
+
+        if module.check_mode:
+            module.exit_json(changed=_system_state_change(module, router))
+
+        if state == 'present':
+            if not router:
+                router = cloud.create_router(name, admin_state_up)
+                module.exit_json(changed=True, result="created",
+                                 id=router['id'])
+            else:
+                if _needs_update(router, admin_state_up):
+                    cloud.update_router(router['id'],
+                                        admin_state_up=admin_state_up)
+                    module.exit_json(changed=True, result="updated",
+                                     id=router['id'])
+                else:
+                    module.exit_json(changed=False, result="success",
+                                     id=router['id'])
+
+        elif state == 'absent':
+            if not router:
+                module.exit_json(changed=False, result="success")
+            else:
+                cloud.delete_router(name)
+                module.exit_json(changed=True, result="deleted")
+
+    except shade.OpenStackCloudException as e:
+        module.fail_json(msg=e.message)
+
+# this is magic, see lib/ansible/module_common.py
+from ansible.module_utils.basic import *
+from ansible.module_utils.openstack import *
+main()
```

Comment:
I think we want to do if __name__ blah blah here for getting the main function to run.